src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/DataSourceParser.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient.SNI;
+using Microsoft.Data.SqlClientX.Handlers;
+
+namespace Microsoft.Data.SqlClient.Microsoft.Data.SqlClientX.Handlers
+{
+    /// <summary>
+    /// Handler to populate data source information required for transport handler
+    /// </summary>
+    internal class DataSourceParser : IHandler<ConnectionHandlerContext>
+    {
+        public IHandler<ConnectionHandlerContext> NextHandler { get; set; }
+
+        public async ValueTask Handle(ConnectionHandlerContext request, bool isAsync, CancellationToken ct)
+        {
+            SqlConnectionString result = new SqlConnectionString(request.connectionString);
+            ServerInfo serverInfo = new ServerInfo(result);
+            string fullServerName = serverInfo.ExtendedServerName;
+            string localDBDataSource = GetLocalDBDataSource(fullServerName, out bool errorWithLocalDBProcessing);
+
+            if (errorWithLocalDBProcessing)
+            {
+                request.error = new Exception(localDBDataSource);
+                return;
+            }
+            
+            // If a localDB Data source is available, we need to use it.
+            fullServerName = localDBDataSource ?? fullServerName;
+            DataSource details = DataSource.ParseServerName(fullServerName);
+            if (details == null)
+            {
+                request.error = new Exception(localDBDataSource);
+                return;
+            }
+            
+            request.dataSource = details;
+            
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(request, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private static string GetLocalDBDataSource(string fullServerName, out bool error)
```

Comment:
Two thoughts here:
* I don't think we should be returning an error flag via an out argument. Since all the error conditions in this method return `null`, can we just have `null` indicate that there's an error? Or can we just throw from here?
* If we cannot eliminate the error flag, I think the more common pattern for methods with error flags is to return bool and have the value returned via an out param. This makes it easier to use it as the condition of an `if` statement. Eg `private static bool TryGetLocalDbDataSource(string fullServerName, out string localDbDataSource)`

Reply:
Ack. I added it as a TODO, will refactor when updating the async part.